src/Core/Algorithms/Math/AppendMatrix.cc

```diff
@@ -24,68 +24,120 @@
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.
+   
+   Author : Moritz Dannhauer (reimplementation)
+   Date:   August 2017
 */
 
 #include <Core/Algorithms/Math/AppendMatrix.h>
 #include <Core/Datatypes/DenseMatrix.h>
+#include <Core/Datatypes/SparseRowMatrix.h>
+#include <Core/Datatypes/SparseRowMatrixFromMap.h>
+#include <Core/Datatypes/Legacy/Matrix/MatrixOperations.h>
 #include <Core/Algorithms/Base/AlgorithmVariableNames.h>
+#include <Core/Datatypes/MatrixTypeConversions.h>
+#include <Eigen/Sparse>
 
 using namespace SCIRun::Core::Algorithms;
 using namespace SCIRun::Core::Algorithms::Math;
 using namespace SCIRun::Core::Datatypes;
 
+const AlgorithmInputName AppendMatrixAlgorithm::InputMatrices("InputMatrices");
+
 AppendMatrixAlgorithm::AppendMatrixAlgorithm()
 {
   addParameter(Variables::RowsOrColumns, 0);
 }
 
-AppendMatrixAlgorithm::Outputs AppendMatrixAlgorithm::run(const AppendMatrixAlgorithm::Inputs& input, const AppendMatrixAlgorithm::Parameters& params) const
+
+bool AppendMatrixAlgorithm::check_dimensions(const Matrix& mat1, const Matrix& mat2, const AppendMatrixAlgorithm::Parameters& params) const
 {
-  DenseMatrixConstHandle lhsPtr = input.get<0>();
-  DenseMatrixConstHandle rhsPtr = input.get<1>();
-  if (!lhsPtr || !rhsPtr)
-    return Outputs(); /// @todo: error
+ auto rows_m1=mat1.nrows(),rows_m2=mat2.nrows(),cols_m1=mat1.ncols(),cols_m2=mat2.ncols();
+ if (params == ROWS)
+ {
+  if(cols_m1!=cols_m2) return false;
+ } else
+  if(rows_m1!=rows_m2) return false;
+ 
+ return true;
+}
+
+AppendMatrixAlgorithm::Outputs AppendMatrixAlgorithm::ConcatenateMatrices(const MatrixHandle base_matrix, const std::vector<boost::shared_ptr<Matrix>> input_matrices, const AppendMatrixAlgorithm::Parameters& params) const
+{  
 
-  const DenseMatrix& lhs = *lhsPtr;
-  const DenseMatrix& rhs = *rhsPtr;
+ if (input_matrices.size()==0)
+   return base_matrix;
+   
+ Outputs outputs = run(boost::make_tuple(base_matrix, input_matrices[0]), params);
+ for(Eigen::Index c=1; c<input_matrices.size(); c++)
+ {
+  auto concatenated = run(boost::make_tuple(outputs, input_matrices[c]), params);
+  outputs=concatenated;
+ }
+
+ return outputs;
+}
 
-  if (params == ROWS)
+AppendMatrixAlgorithm::Outputs AppendMatrixAlgorithm::run(const AppendMatrixAlgorithm::Inputs& input, const AppendMatrixAlgorithm::Parameters& params) const
+{ 
+  auto lhsPtr = input.get<0>();
+  auto rhsPtr = input.get<1>();
+  if (!lhsPtr || !rhsPtr)
+   error(" At least two matrices are needed to run this module. ");
+
+  if (!((matrixIs::sparse(lhsPtr) && matrixIs::sparse(rhsPtr)) || (matrixIs::dense(lhsPtr) && matrixIs::dense(rhsPtr)) || (matrixIs::column(lhsPtr) && matrixIs::column(rhsPtr))))  
```

Comment:
This seems to be where the issues of the dense column or row matrices are happening